dapp/contracts/loans/InvestorLedger.sol

```diff
@@ -1,158 +1,452 @@
-        require(account.loanToken.transferFrom(account.liege, this, account.totalPaybackNeeded));
+        /// Withdrawal counters. These increase as different types of withdrawals
+        /// are performed.
+        mapping(address => WithdrawalData) withdrawalData;
 
-        collateralToLiege(account);
+        /// Main ledger FSM state.
+        State state;
     }
 
-    function markCancelled(Ledger storage account) public {
-        require(!account.loanCancelled && !account.loanDefaulted);
+    struct InvestorData {
+        uint256 amountInvested;
+    }
+
+    struct WithdrawalData {
+        uint256 loanWithdrawn;
+        uint256 collateralWithdrawn;
+    }
 
-        account.loanCancelled = true;
+    /// View functions for the Ledger structure that calculate denormalized
+    /// data based on investment status.
+    
+    // amountInvested(address) returns how much was invested into this loan by
+    // a particular address.
+    function amountInvested(Ledger storage ledger, address investor) view public returns (uint256 amount) {
+        return ledger.investorData[investor].amountInvested;
+    }
 
-        collateralToLiege(account);
+    // paybackRequired returns how much payback is/will be required for this
+    // loan (loan amount + interest).
+    function paybackRequired(Ledger storage ledger) view public returns (uint256 amount) {
+        return ledger.amountWanted + calculateInterest(ledger, ledger.amountWanted);
     }
 
-    function markDefaulted(Ledger storage account) public {
-        require(!account.loanCancelled && !account.loanDefaulted);
+    // calculateInterest is a convenience function to calculate the interest
+    // of this loan on a given value.
+    function calculateInterest(Ledger storage ledger, uint256 investment) view private returns (uint256 interest) {
+        return investment * ledger.interestPermil / PERMIL;
+    }
+
+    // openAccount is a static constructor of the Ledger state object.
+    // @param collateralToken: Token to be used as collateral during the loan.
+    // @param loanToken: Token to be loaned.
+    // @param borrower: Receiver of the loan.
+    // @param totalLoanNeded: How much loanToken does borrower want to borrow.
+    // @param interestPermil: Loan interest, in permil (1/10th of a percent).
+    function openAccount(
+        IToken collateralToken,
+        IToken loanToken,
+        address borrower,
+        uint256 amountWanted,
+        uint16 interestPermil,
+        uint64 fundraisingDelta,
+        uint64 paybackDelta) internal view returns (Ledger account)
+    {
+        // Argument validation.
+        require(fundraisingDelta > 0);
+        require(paybackDelta > 0);
+        require(amountWanted > 0);
+        require(block.timestamp + fundraisingDelta + paybackDelta > block.timestamp);
+        require(block.timestamp + fundraisingDelta + paybackDelta > block.timestamp + fundraisingDelta);
```

Comment:
Nevermind, this IS for math overflows

Reply:
201 is actually an overflow check - checks that the nearest possible sum of all deltas and now() do not end up in the past.  202 is also an overflow check IIRC, but I can't for the life of me understand what I meant when I wrote it. Since we're both confused, it means it needs more comments :).

If you think these are invalid, I'll try to reason again about this and write better ones. The purpose of them being here is so that the contract doesn't get stuck later on in a transition if the timestamps were crafted maliciously.

Reply:
Alright. I'll add comments then :)